Optimize import_openssl.sh, openssl.cmake and openssl.config

Admincore-AppDev opened the pull request.

schwabe commented on import_openssl.sh:
```diff
@@ -256,16 +275,13 @@ function generate_build_config_headers() {
-  make include/openssl/core_names.h include/internal/param_names.h crypto/params_idx.c
-  make include/openssl/comp.h include/openssl/x509_acert.h
+  # Refactored to this, because it search for all existing *.in files to transform it in *.h and *.c files.
+  # There is no need to search and declare them manually anymore, now it's search automatically.
```
This belongs into the commit message. Keeping that comment forever in the file is not a good idea.

schwabe commented on import_openssl.sh:
```diff
@@ -256,16 +275,13 @@ function generate_build_config_headers() {
+  # Refactored to this, because it search for all existing *.in files to transform it in *.h and *.c files.
+  # There is no need to search and declare them manually anymore, now it's search automatically.
+  local list_existing_in_files=$(find crypto include providers -name '*.in' | awk '{gsub(/\.in/, ""); print}' | sort -u)
+  echo -e "Founded *.in files are:\n$list_existing_in_files\n"
```
Found instead Founded

schwabe commented on import_openssl.sh:
```diff
@@ -35,9 +35,28 @@ export LANG=C
-    echo ANDROID_NDK_HOME not set to a valid directory
-    exit 1
+# Declare Android-NDK ABI-Level
+NDK_TOOLCHAIN_ABILVL=28
```
I think SDKINT instead of ABILVL is more normal way of expressing this in Android land.

schwabe commented on import_openssl.sh:
```diff
@@ -256,16 +275,13 @@ function generate_build_config_headers() {
+  for input_in_files in "${list_existing_in_files[@]}"; do make $input_in_files; done
 
-  rm -f apps/CA.pl.bak openssl/opensslconf.h.bak
+  #rm -f apps/CA.pl.bak openssl/opensslconf.h.bak # Comment out, because these files are not exist anymore.
```
Just remove it completely instead of commenting it out. The reason of change belong in the commit message.

Admincore-AppDev commented:
Thanks for reviewing @schwabe, after this was my first commit ever i have still to learn more about commits. Comments to code i should do only with comments in github or in commit descrption, not in files - thanks for this lesson.

Can you besides tell me something. You are using this script to update this repo, or you are using an alternative script? You machine is a MacOS right? There is a $var which blows the script on Linux, i already isolated it and have a kind of fix. I only need verificate if that script usually works on your machine, thanks! If i am sure i will make a pull requst and ask you to review. I hope i can fix it until release date of OpenSSL 3.5.0 in April, besides i have a working build already.

schwabe commented:
Yes, I use that script. It is quite brittle but it always good enough for me. One bug that is in this script that it generates NO_ASM defines in the config header files, which I manually fix since I never had the motivation to deep dive what is causing this.

Admincore-AppDev commented:
Thanks for your response!

I can approve that your script adds "NO_ASM" to the configuration headers. I add a routine to remove these defines in configuration headers without changing anything.

I read the [OpenSSL GitHub-Page](https://github.com/openssl/openssl/blob/master/Configurations/10-main.conf) and see what is usually cause this:

```
    "linux-generic32" => {
        inherit_from     => [ "BASE_unix" ],
        CC               => "gcc",
        CXX              => "g++",
        CFLAGS           => picker(default => "-Wall",
                                   debug   => "-O0 -g",
                                   release => "-O3"),
        CXXFLAGS         => picker(default => "-Wall",
                                   debug   => "-O0 -g",
                                   release => "-O3"),
        cflags           => threads("-pthread"),
        cxxflags         => combine("-std=c++11", threads("-pthread")),
        lib_cppflags     => "-DOPENSSL_USE_NODELETE",
        ex_libs          => add("-ldl", threads("-pthread")),
        bn_ops           => "BN_LLONG RC4_CHAR",
        thread_scheme    => "pthreads",
        dso_scheme       => "dlfcn",
        shared_target    => "linux-shared",
        shared_cflag     => "-fPIC",
        shared_ldflag    => sub { $disabled{pinshared} ? () : "-Wl,-znodelete" },
        enable           => [ "afalgeng" ],
    },
```

The alternative could be this:

```
    "linux-x86" => {
        inherit_from     => [ "linux-generic32" ],
        CFLAGS           => add(picker(release => "-fomit-frame-pointer")),
        cflags           => add("-m32"),
        cxxflags         => add("-m32"),
        lib_cppflags     => add("-DL_ENDIAN"),
        bn_ops           => "BN_LLONG",
        asm_arch         => 'x86',
        perlasm_scheme   => "elf",
    },
```
or this:
```
    "android-x86" => {
        inherit_from     => [ "android" ],
        CFLAGS           => add(picker(release => "-fomit-frame-pointer")),
        bn_ops           => add("RC4_INT"),
        asm_arch         => 'x86',
        perlasm_scheme   => "android",
        ex_libs          => add(threads("-latomic")),
    },
```
The linux-generic32/64 defines in openssl.config are not allow to set ASM enabled usually. But after you won't change that to x86/x86_64 for sure, i think this routine i added is usefully.

I try to merge the conflicts caused by import of 3.5.0 and finish my Pull-Request, i made heavy test and be happy.
I did verificated the configuration headers against yours and there is the same.

BTW thanks for importing OpenSSL 3.5.0-beta1, this was helpful while merging.

I hope we can finish this before finale release of OpenSSL 3.5.0 any maybe with next App release? If you have any concerns, please tell me, i am really sure you can have a benefit from that. And i know this is only a side project of you and your focus is on something else. In worst case i will keep this in my fork only and ask the GitHub Support to remove my Pull-Request here.

Admincore-AppDev committed:
Optimize import_openssl.sh, openssl.cmake and openssl.config

Bugfix:

(import_openssl.sh) Fix misspelling with CC= and aarch64 "remove eabi"
(import_openssl.sh) Linux: Fix a script error caused from an empty $var in "generate_build_config_headers", add a checkup to test if $var is empty
(openssl.config)    Linux: Add "no-afalgeng" and "no-padlockeng" to fix compile errors, caused from missing Kernel-Engines which Android can't use

Feature:

(import_openssl.sh) Add Android-NDK check routine for all platforms
(import_openssl.sh) Add loop routine which transforms *.in files into *.h and *.c files, also skip make if a header files has no configuration
(import_openssl.sh) Add routine to remove "NO_ASM" defines in Config-Headers

Additional:

(import_openssl.sh) Clean up some whitespace and empty lines
(openssl.cmake)     Clean up some whitespace, duplicates and empty lines
(openssl.cmake)     Removed unneeded "*.h *.info *.txt" files
(elf)               Delete this empty unneeded file

Info:

Changes were verified against import of OpenSSL 3.5.0 and compiling against ics-openvpn v0.7.58

schwabe commented:
For your commit, do you want to stick with the github author name (From: gq5q92t724dctp494wy
 <139226997+gq5q92t724dctp494wy@users.noreply.github.com>) or do you want a real username/author in your commits?  (like I have: Arne Schwabe <arne@rfc2549.org>)

Admincore-AppDev commented:
Good evening @schwabe,
you got a good point there. This account was never pretend to make commits, so i just created a random name to make bug reports only. After i will contribute now, i have indeed to change my odd username. The email should stay on GitHub-NoReply as filter, but the username is true not good. I have already change it and did again a force push, looks very cluttery, sorry for that. It weas a good experience and i learned a lot about commits, my way was totally wrong, i have to be better in future. For the "*.in" files loop i have made a tiny adapt. to prevent a failure if we have a adapted configuration (like i already used for my local build). OpenSSL 3.5.0 works pretty well with your App, no doubts to make it in release soon. Thanks, your App is a beast!

schwabe commented:
No worries. You are already ding quite good. I am just pointing out things you might miss as beginner. You can even see the user/email on github if you add .patch to the url e.g. https://github.com/schwabe/platform_external_openssl/commit/7cb0aa58f8c33f1a0c3c7e2e6a88437339c7471c.patch

If you are happy with that what is now, then I will just merge the commit tommorrow or the day after then import OpenSSL 3.5.0. 

An easy way to change the git author to what you want on command line git is `git -c user.name="Arne Schwabe" -c user.email=arne@rfc2549.org commit --amend --reset-author`.

Admincore-AppDev commented:
Yes, and I am very thankful for your helping and your advises. The tip with the amend of the commit author is a useful way to do, after I did a more complicated way with rebase - just keep things simple. I already changed the author locally on my machine so it should be there on next commits too. I really thought it is glued to my GitHub account, but seems like I can change the author independently whenever I want.

I am totally happy with the author and can give you my GO to merge. Ofc there would be more space to optimize in future, but I think make the script working on "all platforms" is good enough for the moment. After you never had trouble on your MacOS it's hard to judge from a LinuxOS view, seems like the shell on MacOS works different.

The .patch way on the link is a really nice feature, I didn't known about this. Seems even nice to patch it from there for side testing without alternating the local repo. Thanks!
